### PR TITLE
Add sort listings functionality

### PR DESCRIPTION
**Pull request by taylor-yamashita**
Working with @ysl-ee to add sorting dropdown to "Items Requested" and "Items for Sale" gallery pages. (Taylor on frontend, Youngseo on backend)

Example demos: 

1. On the "Items Requested" page, user selects "price: low to high":
Screen Shot 2023-03-21 at 12 13 56 PM

Result:
Screen Shot 2023-03-21 at 12 14 12 PM

2. On the "Items for Sale" page,  user selects "date posted: newest":
Screen Shot 2023-03-21 at 12 15 29 PM

Result:
Screen Shot 2023-03-21 at 12 15 40 PM

### CODE REVIEW

**Comment by ysl-ee** [comment]
- discovered that getItemsRelative() and getItemRequestsRelative() in views.py do not render directly do a page — they return a json reponse. we will probably have to return to these functions for the actual sorting functionality? we considered using a "sort_type" parameter in request.GET() so that we don't have to have 4 different URLs going on, but we removed it for the moment. (might bring back)
- gallery.html has updateQuery() which updates the query whenever new checkboxes (categories & conditions) are clicked. we are going to mimic this structure and implement the sorting right before it renders to prevent these selections from accidentally going away when we sort.
- category_pks and condition_indexes (from which the active ones are extracted) are declared at the top of the script 
- eventually, getItems() is kind of called (??) (i think and hope??) with a query string (see line 498 in populateItemsSynchronously())

**Comment by taylor-yamashita** [comment]
^ Note: Generally, duplicate code is used to display available items vs. requested items, so we also put duplicate code in their respective places.

**Comment by taylor-yamashita** [comment]
At the moment, everything is connected, but the sorting functions don't appear to behave as expected—e.g. sorting "price low to high" shows items with highest price first. @ysl-ee would you mind looking into this?

**Comment by ysl-ee** [comment]
@taylor-yamashita sortings should be happening in correct order now :)

File: TRT-django/marketplace/templates/marketplace/gallery.html
Comment (nicholaspad) [inline TRT-django/marketplace/templates/marketplace/gallery.html:80]: same nit about the spacing between class names

File: TRT-django/marketplace/templates/marketplace/gallery.html
Comment (nicholaspad) [inline TRT-django/marketplace/templates/marketplace/gallery.html:208]: same nit about setting `sort_type`

File: TRT-django/marketplace/views.py
Comment (nicholaspad) [inline TRT-django/marketplace/views.py:479]: could you instead use this conditional block only to set the value passed into the `order_by` argument, and then make one call to `items.annotate` at the end?

```python
if sort_type == "price_hightolow":
    order_by = [F("price").asc(), F("pk").asc()]
elif ...:
    order_by = ...
elif ...:
    order_by = ...
elif ...:
    order_by = ...
else: ...
    order_by = ...

    # annotate items by search string rank
    items = items.annotate(rank=SearchRank(SearchVector("name", "description"), SearchQuery(search_string), cover_density=True))

items = items.annotate(
            row=Window(
                expression=RowNumber(),
                order_by=order_by, # also order by unique pk to make tie-breaks consistent
            )
        )
```

File: TRT-django/marketplace/views.py
Comment (nicholaspad) [inline TRT-django/marketplace/views.py:1640]: same comment about shortening the conditional block

**Review by nicholaspad — CHANGES_REQUESTED** [review]
awesome work @taylor-yamashita @ysl-ee, i would've found this one really difficult because the retail codebase is quite confusing

just a couple suggestions, but thanks for making this change!! a ton of people wanted this feature :)

File: TRT-django/marketplace/templates/marketplace/browse_item_requests.html
Comment (shannon-heh) [inline TRT-django/marketplace/templates/marketplace/browse_item_requests.html:85]: good touch with keeping consistent with the lowercase :)

**Comment by shannon-heh** [comment]
@ysl-ee @taylor-yamashita could one of you upload a short screen recording or a series of screenshots to demonstrate the working functionality? i'd also like to see what the drop-down looks like

**Review by shannon-heh — CHANGES_REQUESTED** [review]
small nit: could you change cursor behavior when you hover a dropdown item? right now it's the default link cursor, but `cursor: pointer;` would be more appropriate: https://developer.mozilla.org/en-US/docs/Web/CSS/cursor.

besides this and a comment about modularity, the code looks great! i tested the functionality on the dev app and it works well :) awesome job

**Comment by ysl-ee** [comment]
I worked on some minor details (spacing in code, and simplifying the sort_by variable update) and realized that the sort was not working on item requests (D:) and addressed that.

Currently still need to work on:
- Cursor should change to pointer on dropdown
- Using helper methods to simplify conditional block -> I did move things around to the helper methods and they do compile and run, but gets stuck. The most recent commit has the calls to the helper methods commented out, so the website should function.

@taylor-yamashita

File: TRT-django/marketplace/views.py
Comment (Danielpark1239) [reply TRT-django/marketplace/views.py]: ^ I'd also like to see the duplicate logic for items and item requests combined in some way

**Comment by Danielpark1239** [comment]
There was an issue where the price sorting for item requests was going the wrong way, but after redeploying on the dev app, it seems to be working correctly.

Overall, this is a great feature and I hope it makes it easy to add further sorts down the line (ex. by quality or expiration date)! Looking forward to approving once the requested changes are made.

**Review by Danielpark1239 — CHANGES_REQUESTED** [review]
same comments as Nick and Shannon, nothing new

**Comment by taylor-yamashita** [comment]
> small nit: could you change cursor behavior when you hover a dropdown item? right now it's the default link cursor, but `cursor: pointer;` would be more appropriate: https://developer.mozilla.org/en-US/docs/Web/CSS/cursor.
> 
> besides this and a comment about modularity, the code looks great! i tested the functionality on the dev app and it works well :) awesome job

resolved! the other dropdown menus seem to inherit this property from css .btn, so added that for consistency. there is no specific css styling that targets dropdown menus; it all seems to be handled by bootstrap classes and directly in the html.

**Comment by taylor-yamashita** [comment]
> I worked on some minor details (spacing in code, and simplifying the sort_by variable update) and realized that the sort was not working on item requests (D:) and addressed that.
> 
> Currently still need to work on:
> 
> * Cursor should change to pointer on dropdown
> * Using helper methods to simplify conditional block -> I did move things around to the helper methods and they do compile and run, but gets stuck. The most recent commit has the calls to the helper methods commented out, so the website should function.
> 
> @taylor-yamashita

will probably need to sync with @ysl-ee on the second one.

**Comment by taylor-yamashita** [comment]
> @ysl-ee @taylor-yamashita could one of you upload a short screen recording or a series of screenshots to demonstrate the working functionality? for archival purposes :D

Added screenshots to the main PR comment! Let me know if you'd like it to be more detailed.

**Comment by taylor-yamashita** [comment]
Condensed the "if" blocks in getItemsRelative() and getItemRequestsRelative() as per @nicholaspad 's suggestion. I looked at @ysl-ee 's helper methods (sortMyItems and ordering), moved the ordering() code back to the original methods (for now), and deleted duplicate code so the app doesn't break right now.

I know @ysl-ee was trying to make one helper function called by both getItemsRelative() and getItemRequestsRelative(), as per @shannon-heh 's suggestion, but honestly getItemsRelative() and getItemRequestsRelative() are themselves basically identical, so I think if we refactor this, we might want to consider combining those two whole functions into something more general instead of making a bunch of smaller helper functions. I haven't looked into it too much but it seems like the only difference is in variable names/constructing queries. 

If we do decide to refactor getItemsRelative() and getItemRequestsRelative(), I think it would be best for us to put that in a whole separate PR in the future, since there could be a lot of refactoring throughout the backend ("Items for Sale" and "Items Requested" pages mostly work the same, but I think there is a lot of separate duplicate code between them right now). Welcoming any thoughts!

**Review by nicholaspad — APPROVED** [review]
> If we do decide to refactor getItemsRelative() and getItemRequestsRelative(), I think it would be best for us to put that in a whole separate refactoring PR in the future, since there could be a lot of refactoring throughout the backend ("Items for Sale" and "Items Requested" pages mostly work the same, but I think there is a lot of separate duplicate code between them right now).

Completely agree with you about doing a larger refactoring in a future PR since `getItemsRelative` and `getItemRequestsRelative` are almost entirely the same. Thanks for adding this feature!! 🎉 

I'd wait for @shannon-heh or @Danielpark1239 to give one more approval before you merge this!

**Comment by Danielpark1239** [comment]
> Condensed the "if" blocks in getItemsRelative() and getItemRequestsRelative() as per @nicholaspad 's suggestion. I looked at @ysl-ee 's helper methods (sortMyItems and ordering), moved the ordering() code back to the original methods (for now), and deleted duplicate code so the app doesn't break right now.
> 
> I know @ysl-ee was trying to make one helper function called by both getItemsRelative() and getItemRequestsRelative(), as per @shannon-heh 's suggestion, but honestly getItemsRelative() and getItemRequestsRelative() are themselves basically identical, so I think if we refactor this, we might want to consider combining those two whole functions into something more general instead of making a bunch of smaller helper functions. I haven't looked into it too much but it seems like the only difference is in variable names/constructing queries.
> 
> If we do decide to refactor getItemsRelative() and getItemRequestsRelative(), I think it would be best for us to put that in a whole separate refactoring PR in the future, since there could be a lot of refactoring throughout the backend ("Items for Sale" and "Items Requested" pages mostly work the same, but I think there is a lot of separate duplicate code between them right now). Welcoming any thoughts!

I also agree that there are a lot of places where refactoring could help simplify the codebase, but that's not an urgent issue. Thanks for the changes! They look good to me, and I think the sort feature's ready to be merged. 👍

**Review by shannon-heh — APPROVED** [review]
awesome job!!! thanks for making the changes. i'll merge in the PR rn

**Pull request merged**
